packages/dd-trace/src/debugger/devtools_client/condition.js

```diff
@@ -0,0 +1,237 @@
+    switch (type) {
+      case 'eq': return `(${args[0]}) === (${args[1]})`
+      case 'ne': return `(${args[0]}) !== (${args[1]})`
+      case 'gt': return `${guardAgainstCoercionSideEffects(args[0])} > ${guardAgainstCoercionSideEffects(args[1])}`
+      case 'ge': return `${guardAgainstCoercionSideEffects(args[0])} >= ${guardAgainstCoercionSideEffects(args[1])}`
+      case 'lt': return `${guardAgainstCoercionSideEffects(args[0])} < ${guardAgainstCoercionSideEffects(args[1])}`
+      case 'le': return `${guardAgainstCoercionSideEffects(args[0])} <= ${guardAgainstCoercionSideEffects(args[1])}`
+      case 'any': return iterateOn('some', ...args)
+      case 'all': return iterateOn('every', ...args)
+      case 'and': return `(${args.join(') && (')})`
+      case 'or': return `(${args.join(') || (')})`
+      case 'startsWith': return `String.prototype.startsWith.call(${assertString(args[0])}, ${assertString(args[1])})`
+      case 'endsWith': return `String.prototype.endsWith.call(${assertString(args[0])}, ${assertString(args[1])})`
+      case 'contains': return `((obj, elm) => {
+          if (${isString('obj')}) {
+            return String.prototype.includes.call(obj, elm)
+          } else if (Array.isArray(obj)) {
+            return Array.prototype.includes.call(obj, elm)
+          } else if (obj instanceof Object.getPrototypeOf(Int8Array)) {
+            return Object.getPrototypeOf(Int8Array.prototype).includes.call(obj, elm)
+          } else if (obj instanceof Set) {
+            return Set.prototype.has.call(obj, elm)
+          } else if (obj instanceof WeakSet) {
+            return WeakSet.prototype.has.call(obj, elm)
+          } else if (obj instanceof Map) {
+            return Map.prototype.has.call(obj, elm)
+          } else if (obj instanceof WeakMap) {
+            return WeakMap.prototype.has.call(obj, elm)
+          } else {
+            throw new TypeError('Variable does not support contains')
+          }
+        })(${args[0]}, ${args[1]})`
+      case 'matches': return `((str, regex) => {
+          if (${isString('str')}) {
+            if (Object.getPrototypeOf(regex) === RegExp.prototype) {
+              return RegExp.prototype.test.call(regex, str)
+            } else if (${isString('regex')}) {
+              return String.prototype.match.call(str, regex) !== null
+            } else {
+              throw new TypeError('Regular expression must be either a string or an instance of RegExp')
+            }
+          } else {
+            throw new TypeError('Variable is not a string')
+          }
+        })(${args[0]}, ${args[1]})`
+      case 'filter': return `(($dd_var) => {
+          return ${isIterableCollection('$dd_var')}
+            ? Array.from($dd_var).filter(($dd_it) => ${args[1]})
+            : Object.entries($dd_var).filter(([$dd_key, $dd_value]) => ${args[1]}).reduce((acc, [k, v]) => {
+                acc[k] = v
+                return acc
+              }, {})
```

Comment:
Nit: the filter could be done in the reduce. That way there is one less intermediate array allocation.

Reply:
Fixed 👍